### DISCUSSION TITLE
Java Loader: Create Function with Custom Types?

### DISCUSSION THREAD

**Discussion by slinga-homebrew** [NationalSecurityAgency/ghidra#5098] [Q&A] [answered]
I would like to define a function with custom data types within the loader. The code I am modifying already creates a function:

```
program.getFunctionManager().createFunction("some_bios_func", addr, addrSet, SourceType.IMPORTED);
```
createFunction() returns a ```Function``` object on success. It looks like I need to call ```Function.updateFunction​()``` but I am unclear on the syntax.

1) Can you point me at example of how to call updateFunction()? 
2) How do I handle the case where one of the arguments is a custom struct? Presumably I need to define the struct before hand and add it to a type lib somehow. Can you also provide an example of that?

Thanks in advance.

**Comment by ghidra1** [comment, marked as answer]
Using a development setup with Eclipse and our full source (i.e., git clones), you can search for all uses of the `Function.updateFunction` method.   There are quite a few.  Here is an modified example from one of our test cases:
```
 	Function function =
		p.getFunctionManager()
				.createFunction(name, addr(functionEntry),
					new AddressSet(addr(functionEntry), addr(functionEntry + size - 1)),
					SourceType.USER_DEFINED);
	ReturnParameterImpl returnParam =
		new ReturnParameterImpl(IntegerDataType.dataType, program);
	StructureDataType struct = new StructureDataType("Struct1", 0, p.getDataTypeManager());
	struct.add(ByteDataType.dataType, "a", null);
	struct.add(ByteDataType.dataType, "b", null);
	ParameterImpl param1 = new ParameterImpl("p1", new PointerDataType(struct), program);
	ParameterImpl param2 = new ParameterImpl("p2", IntegerDataType.dataType, program);
	function.updateFunction(
		program.getCompilerSpec().getDefaultCallingConvention().getName(), returnParam,
		FunctionUpdateType.DYNAMIC_STORAGE_FORMAL_PARAMS, true, SourceType.USER_DEFINED,
		param1, param2);
	return function;
```
Any datatype used in defining a parameter/return will be resolved when the update is completed.  If the datatype does not already exist you can certainly build it up and use.  Be sure to specify appropriate DataTypeManager when building a structure so that the correct data organization is used when adding components.
